app/controlplane/internal/usercontext/currentorganization_middleware.go

```diff
@@ -0,0 +1,104 @@
+//
+// Copyright 2024 The Chainloop Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package usercontext
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	v1 "github.com/chainloop-dev/chainloop/app/controlplane/api/controlplane/v1"
+	"github.com/chainloop-dev/chainloop/app/controlplane/pkg/biz"
+	"github.com/go-kratos/kratos/v2/log"
+	"github.com/go-kratos/kratos/v2/middleware"
+)
+
+type Org struct {
+	ID, Name  string
+	CreatedAt *time.Time
+}
+
+func WithCurrentOrg(ctx context.Context, org *Org) context.Context {
+	return context.WithValue(ctx, currentOrgCtxKey{}, org)
+}
+
+func CurrentOrg(ctx context.Context) *Org {
+	res := ctx.Value(currentOrgCtxKey{})
+	if res == nil {
+		return nil
+	}
+	return res.(*Org)
+}
+
+type currentOrgCtxKey struct{}
+
+func WithCurrentOrganizationMiddleware(userUseCase biz.UserOrgFinder, logger *log.Helper) middleware.Middleware {
+	return func(handler middleware.Handler) middleware.Handler {
+		return func(ctx context.Context, req interface{}) (interface{}, error) {
+			// Get the current user and return if not found, meaning we are probably coming from an API Token
+			u := CurrentUser(ctx)
+			if u == nil {
+				return handler(ctx, req)
+			}
+
+			var err error
+			ctx, err = setCurrentOrganization(ctx, userUseCase, logger)
+			if err != nil {
+				return nil, fmt.Errorf("error setting current org: %w", err)
+			}
+
+			org := CurrentOrg(ctx)
+			if org == nil {
+				return nil, errors.New("org not found")
+			}
+
+			logger.Infow("msg", "[authN] processed organization", "org-id", org.ID, "credentials type", "user")
+
+			return handler(ctx, req)
+		}
+	}
+}
+
+// Find the current membership of the user and sets it on the context
+func setCurrentOrganization(ctx context.Context, userUC biz.UserOrgFinder, logger *log.Helper) (context.Context, error) {
+	u := CurrentUser(ctx)
```

Comment:
This check is done from the caller, not sure if we can omit it here (or there)

Reply:
We can probably remove this one, the other one is needed.